keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -1635,7 +1641,7 @@ conditions apply:
   Filter
 
 One of the ResourceClaims satisfying these criteria is picked randomly and deallocation
-is requested by setting the Deallocate field. The scheduler then needs to wait
+is requested by setting the ResourceClaimStatus.DeallocationRequested field. The scheduler then needs to wait
```

Comment:
I still think it's better if the driver decides for itself when to deallocate.

For example, if it's a local resource, it might never deallocate from a node. Otherwise, it could start deallocating from a node when reservedFor is empty, perhaps even with a grace period.

Now, if we add a grace period, kube-scheduler might want to give a signal that it "thinks" that a deallocation could help. This signal could work this way:

- the scheduler clears the `selectedNode` field.
- the driver observes:
  - there is a Pod with an empty PodScheduling.spec.selectedNode that requires this allocated claim. Note that this is different from a Pod that was never attempted for scheduling because the PodScheduling object wouldn't exist at this time.
  - the claim is not reserved for any other pod

Then the driver could choose to trigger deallocation.

Two advantages:
- The scheduler only updates one object.
- The driver could be more pro-active.

Reply:
> I still think it's better if the driver decides for itself when to deallocate.
>
> For example, if it's a local resource, it might never deallocate from a node. Otherwise, it could start deallocating from a node when reservedFor is empty, perhaps even with a grace period.

A driver should never deallocate a resource that has a non-empty reservedFor. It cannot be sure that the resource is not in use or about to be used.

To make this work, the scheduler would first have to clear the reservedFor to indicate that deallocating the claim is okay. That's doable.

But that alone puts the claim into "allocated, available" state. That's not a state where the driver pro-actively frees the claim because it doesn't know whether the allocation is still needed. For example, claims with immediate allocation enter that state before a pod gets created which uses them.

So we need some additional indicator.

> Now, if we add a grace period, kube-scheduler might want to give a signal that it "thinks" that a deallocation could help. 
> This signal could work this way:
> 
>     * the scheduler clears the `selectedNode` field.
> 
>     * the driver observes:
>       
>       * there is a Pod with an empty PodScheduling.spec.selectedNode that requires this allocated claim. Note that this is different from a Pod that was never attempted for scheduling because the PodScheduling object wouldn't exist at this time.
>       * the claim is not reserved for any other pod
> 
> 
> Then the driver could choose to trigger deallocation.

That would work, but there's a race condition:
- The driver observes the above state and starts deallocation.
- Another pod gets created which can use the same claim.
- The pod gets added to reservedFor and gets scheduled to a node.
- The driver finishes the deallocation and the resource is not available for the pod anymore.

This race can be fixed by adding a `ClaimStatus.Deallocating` boolean that the driver would have to set before starting the deallocation.

I think that would work. Setting `reservedFor` together with `allocation` becomes more important, but that's already done. Does it sound better to you?

> Two advantages:
> 
>     * The scheduler only updates one object.

Agreed.

>     * The driver could be more pro-active.

No, I think it still has to be reactive. But even that is an improvement.

Reply:
I can't tell if there was some compromise found here?  How do we end up in this situation?  Is it something like:

* pod with 2 claims
* driver-1 allocates and only available on nodes (a, b, c)
* driver-2 allocates and only available on nodes (x, y, z)
?  That shouldn't happen, right?

I guess that could play out as:

* pod with 2 claims
* scheduler suggests (a, b, c)
* both drivers ACK
* scheduler chooses a
* driver-1 allocates and only available on nodes (a, b, c)
* driver-2 allocates and only available on nodes (a, b, c)
* Pod tries to run and resource 2 is no longer availble on a ?

Help me understand how we end up here?

Reply:
This discussion has been superseded by https://github.com/kubernetes/enhancements/pull/3502#discussion_r984666208 below. After thinking about this some more, I realized that @alculquicondor's proposal depended on not releasing a reservation (because doing so would trigger deallocation, according to the definition above) but not releasing a reservation after a failed pod scheduling attempt leads to deadlocks or poor resource utilization. I gave examples for both below.

To answer your specific example: you are right, both drivers should allocate so that the claims are usable on the same node. The problem is not with where claims are usable, but rather which pod is allowed to use a claim exclusively when claims cannot be shared.

Reply:
I think @alculquicondor accepted my line of argumentation because he added [his LGTM](https://github.com/kubernetes/enhancements/pull/3502#issuecomment-1268130531) after we discussed this.

Reply:
> The problem is not with where claims are usable, but rather which pod is allowed to use a claim exclusively when claims cannot be shared

Not needed for this review, but can you illuminate an example time-sequence that leads to that?